redfish-core/lib/power_supply_metrics.hpp

```diff
@@ -0,0 +1,350 @@
+        [aResp, chassisID, powerSupplyID](
+            const boost::system::error_code ec,
+            const std::vector<std::pair<
+                std::string,
+                std::vector<std::pair<std::string, std::vector<std::string>>>>>&
+                intfSubTree) mutable {
+            if (ec)
+            {
+                BMCWEB_LOG_DEBUG << "D-Bus response error on GetSubTree " << ec;
+                messages::internalError(aResp->res);
+                return;
+            }
+
+            std::string serviceName;
+            std::string averagePath;
+            std::string maximumPath;
+
+            for (const auto& [objectPath, connectionNames] : intfSubTree)
+            {
+                if (objectPath.empty())
+                {
+                    BMCWEB_LOG_DEBUG << "Error getting D-Bus object!";
+                    messages::internalError(aResp->res);
+                    return;
+                }
+
+                const std::string& validPath = objectPath;
+                auto psuMatchStr = powerSupplyID + "_input_power";
+                std::string psuInputPowerStr;
+                // validPath -> {psu}_input_power
+                // 5 below comes from
+                // /org/open_power/sensors/aggregation/per_30s/
+                //   0      1         2         3         4
+                // .../{psu}_input_power/[average,maximum]
+                //           5..............6
+                if (!dbus::utility::getNthStringFromPath(validPath, 5,
+                                                         psuInputPowerStr))
+                {
+                    BMCWEB_LOG_ERROR << "Got invalid path " << validPath;
+                    messages::invalidObject(aResp->res, validPath);
+                    return;
+                }
+
+                if (psuInputPowerStr != psuMatchStr)
+                {
+                    // not this power supply, continue to next objectPath...
+                    continue;
+                }
+
+                BMCWEB_LOG_DEBUG << "Got validPath: " << validPath;
+                for (const auto& connection : connectionNames)
+                {
+                    serviceName = connection.first;
+
+                    for (const auto& interfaceName : connection.second)
+                    {
+                        if (interfaceName == "org.open_power.Sensor."
+                                             "Aggregation.History.Average")
+                        {
+                            averagePath = validPath;
+                        }
+                        else if (interfaceName == "org.open_power.Sensor."
+                                                  "Aggregation.History.Maximum")
+                        {
+                            maximumPath = validPath;
+                        }
+                    }
+                }
+            }
+
+            getAverageMaximumValues(aResp, serviceName, averagePath,
```

Comment:
I would suggest splitting getAverageMaximumValues into 
getAverageValues and getMaximumValues. Then call the respective functions at lines 279 and 284. 
I think for performance (maybe negligible) we should NOT chain the dbus average/maximum dbus calls?

Reply:
Perhaps I misunderstood something from the conversation I had with Shawn. We need the average values and the maximum values to both be completed before we can proceed with parsing the data those retrieved and putting it in the Redfish response. As I understood it, chaining them together was seen as the best way to ensure we had all the data.

Reply:
If you are suggesting calling getAverageValues() and getMaximumValues() in parallel, it makes it difficult to compile the results and emit the JSON at the right time.  You need to wait for both to complete, and it is not clear which one will run first as far as I know.  So doing them serially, while it takes a little longer, will ensure we don't try to compile the results and emit JSON until all the D-Bus calls are done.

Reply:
What Shawn said. :-)

Reply:
As long as we have a reason, I'm good with this. I'm a little disappointed in our json emitting time but better to these value arrays in sync than not.

Reply:
Okay, going to consider this conversation resolved.